### PR TITLE
Report more details of unobtainable ShardLock

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today a common reason for a `ShardLockObtainFailedException` is when a
shard is removed from a node and then assigned straight back to it again
before the node has had a chance to shut the previous shard instance
down. For instance, this can happen if a node briefly leaves the cluster
holding a primary with no in-sync replicas.

The message in this case is typically as follows:

    obtaining shard lock timed out after 5000ms, previous lock details: [shard creation] trying to lock for [shard creation]

This is pretty hard to interpret, and doesn't raise the important
question: "why didn't the shard shut down sooner?"

With this change we reword the message a bit, report the age of the
shard lock, and adjust the details to report that the lock is held by a
closing shard:

    obtaining shard lock for [starting shard] timed out after [5000ms], lock already held for [closing shard] with age [12345ms]

Relates #38807

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -854,17 +860,23 @@ private void decWaitCount() {
         void acquire(long timeoutInMillis, final String details) throws ShardLockObtainFailedException {
             try {
                 if (mutex.tryAcquire(timeoutInMillis, TimeUnit.MILLISECONDS)) {
-                    lockDetails = details;
+                    lockDetails = Tuple.tuple(System.nanoTime(), details);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/env/NodeEnvironment.java:863]: NIT: `setDetails(details);`

**Review by original-brownbear — APPROVED** [review]
LGTM

**Review by dakrone — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Thanks both

**Pull request merged**